ak.array_equal() fails on virtual arrays and returns wrong result for empty unions

*This issue was drafted and posted by Claude Code, based on findings from property-based testing work on PR #3887.*

## Summary

`ak.array_equal()` has two categories of bugs discovered during property-based testing of `to_buffers`/`from_buffers` roundtrips (PR #3887 / branch `sakuma/property-based-testing`):

1. **Exceptions with virtual arrays**: `ak.array_equal()` calls `to_packed()` (via `ak_almost_equal.py:95-96`) which triggers materialization. This crashes for arrays with virtual buffers.
2. **Wrong result with empty union arrays**: `ak.array_equal()` returns `False` for identical empty union arrays.

---

## Bug 1: Exceptions with virtual arrays

`ak.array_equal()` calls `to_packed()` on both inputs, which in turn calls `materialize(VirtualNDArray)._to_packed(True)`. For arrays with virtual/placeholder buffers (e.g., arrays produced by `from_buffers`, or arrays generated by `hypothesis-awkward` with `allow_virtual=True`), this triggers several different crashes depending on the array structure.

Note: `ak.materialize()` and `ak.to_list()` are also affected because they both call `to_packed()` internally.

### Error variants

**a) `TypeError: cannot interpret unknown lengths as concrete values`**

Triggered when shape generators in `from_buffers` attempt `len()` on a `PlaceholderArray` with unknown length.

```
  File "ak_almost_equal.py", line 95, in _impl
    left_layout = layouts[0].to_packed()
  ...
  File "ak_from_buffers.py", line 631, in _adjust_length
    return 0 if len(offsets) == 1 else offsets[-1]
  File "placeholder.py", line 160, in __len__
    return int(self._shape[0])
  File "shape.py", line 82, in __int__
    raise TypeError("cannot interpret unknown lengths as concrete values")
```

Falsifying example: `0 * var * {"": var * (unknown, var * string)}`

**b) `AssertionError: RegularArray length must be an integer`** (and similarly for `RecordArray`)

Triggered when `.length` is accessed during materialization but nested contents still have `unknown_length`.

```
  File "ak_almost_equal.py", line 95, in _impl
    left_layout = layouts[0].to_packed()
  ...
  File "regulararray.py", line 271, in length
    assert is_integer(self._length), (
AssertionError: RegularArray length must be an integer for an array with concrete data, not <class 'awkward._nplikes.shape.UnknownLength'>
```

Falsifying example: `0 * var * var * var * 0 * bool`

The same pattern occurs for `RecordArray`:

```
  File "recordarray.py", line 375, in length
    assert is_integer(self._length), (
AssertionError: RecordArray length must be an integer for an array with concrete data, not <class 'awkward._nplikes.shape.UnknownLength'>
```

Falsifying example: `0 * var * {"": bool}`

**c) `RuntimeError: PlaceholderArray should never have been encountered`**

Triggered when shape generators in `from_buffers` attempt arithmetic on `PlaceholderArray` buffers that should have been replaced during materialization.

```
  File "ak_almost_equal.py", line 95, in _impl
    left_layout = layouts[0].to_packed()
  ...
  File "placeholder.py", line 67, in materialize
    raise RuntimeError(msg)
RuntimeError: PlaceholderArray(...) should never have been encountered.
  Awkward Array tried to access a buffer at 'node1-stops', but it exists only as a placeholder.
```

Falsifying example: `0 * var * var * bool`

**d) `AssertionError: placeholder arrays that are sliced should have known shapes`**

Triggered when `ListArray.to_ListOffsetArray64()` (called from `_to_packed`) tries to slice `stops[:-1]` on a placeholder with unknown length.

```
  File "listarray.py", line 1639, in _to_packed
    return self.to_ListOffsetArray64(True).to_packed(recursive)
  File "listarray.py", line 295, in to_ListOffsetArray64
    if (not nplike.known_data) or nplike.array_equal(starts[1:], stops[:-1]):
  File "placeholder.py", line 111, in __getitem__
    raise AssertionError("placeholder arrays that are sliced should have known shapes")
```

Falsifying example: `0 * var * var * bool`

**e) `TypeError: cannot interpret unknown lengths as concrete index values`**

Triggered when `_to_packed` tries to slice content by offset values that are still unknown.

```
  File "listoffsetarray.py", line 2330, in _to_packed
    next_content = next._content[: next._offsets[-1]]
  ...
  File "shape.py", line 79, in __index__
    raise TypeError("cannot interpret unknown lengths as concrete index values")
```

Falsifying example: `0 * string`

### Reproducer (requires `hypothesis-awkward`)

```python
from hypothesis import given, settings
import hypothesis_awkward.strategies as st_ak
import awkward as ak

@settings(max_examples=200)
@given(a=st_ak.constructors.arrays(allow_virtual=True))
def test_array_equal_virtual(a):
    ak.array_equal(a, a)  # crashes for many array types

test_array_equal_virtual()
# ExceptionGroup: Hypothesis found 6 distinct failures.
```

---

## Bug 2: Wrong result for identical empty union arrays

`ak.array_equal()` returns `False` when comparing identical empty union arrays. This is a logic error in the union comparison code (`ak_almost_equal.py:249-287`), likely in `ordered_unique_values` which calls `np.unique_all` on an empty tag array.

### Reproducer

```python
import awkward as ak
import numpy as np

form = ak.forms.from_dict({
    "class": "UnionArray",
    "tags": "i8",
    "index": "i64",
    "contents": [
        {
            "class": "RecordArray",
            "fields": None,
            "contents": [
                {"class": "NumpyArray", "primitive": "bool", "form_key": "node2"},
                {"class": "NumpyArray", "primitive": "bool", "form_key": "node3"},
            ],
            "form_key": "node1",
        },
        {"class": "EmptyArray", "form_key": "node4"},
    ],
    "form_key": "node0",
})
container = {
    "node0-tags": np.array([], dtype=np.int8),
    "node0-index": np.array([], dtype=np.int64),
    "node2-data": np.array([], dtype=np.bool_),
    "node3-data": np.array([], dtype=np.bool_),
}
a = ak.from_buffers(form, 0, container)
print(a.type)       # 0 * union[(bool, bool), unknown]
print(ak.array_equal(a, a))  # False  (should be True)
```

---

## Relevant code

- `src/awkward/operations/ak_almost_equal.py` lines 95-96 (`to_packed()` calls) and 249-287 (union comparison logic)
- `src/awkward/contents/content.py` line 1199 (`to_packed` calls `materialize`)
- `src/awkward/_nplikes/placeholder.py` lines 67, 109-113, 160
- `src/awkward/contents/recordarray.py` line 375 (length assertion)
- `src/awkward/contents/regulararray.py` line 271 (length assertion)

## Related

- PR #3887 (property-based testing for `to_buffers`/`from_buffers`) — discovery context
- `tests/test_3773_almost_equal_unionarray.py` — existing union comparison tests
- `tests/test_3637_skip_placeholder_materialization_in_to_packed.py` — existing placeholder tests

--- a/src/awkward/operations/ak_almost_equal.py
+++ b/src/awkward/operations/ak_almost_equal.py
@@ -144,6 +144,9 @@ def _impl(
                 mergecastable="equiv" if dtype_exact else "family",
                 dropunused=True,
             )
+            # UnionArray simplifications can produce IndexedArrays
+            if left.is_indexed and not left.is_option:
+                left = left.project()
         if right.is_union:
             right = right.simplified(
                 right.tags,
@@ -154,6 +157,9 @@ def _impl(
                 mergecastable="equiv" if dtype_exact else "family",
                 dropunused=True,
             )
+            # UnionArray simplifications can produce IndexedArrays
+            if right.is_indexed and not right.is_option:
+                right = right.project()
 
         # Simplify regular NumPy types
         if left.is_numpy and left.purelist_depth > 1:

--- a/tests/properties/operations/test_to_from_buffers.py
+++ b/tests/properties/operations/test_to_from_buffers.py
@@ -8,33 +8,10 @@ from hypothesis import given, settings
 import awkward as ak
 
 
-@settings(max_examples=500)
-@given(a=st_ak.constructors.arrays(allow_union=False))
+@settings(max_examples=1000)
+@given(a=st_ak.constructors.arrays())
 def test_roundtrip(a: ak.Array) -> None:
-    """`to_buffers` followed by `from_buffers` reconstructs the array.
-
-    Union arrays are excluded because of known issues:
-
-    - Union arrays: `ak.array_equal` returns `False` for some empty union
-      arrays that are identical, e.g., `0 * union[unknown, (bool, bool)]`.
-
-    See `test_roundtrip_no_equality_check` for a separate test that includes
-    union arrays but only asserts that the roundtrip runs.
-    """
+    """`to_buffers` followed by `from_buffers` reconstructs the array."""
     sent = ak.to_buffers(a)
     returned = ak.from_buffers(*sent)
     assert ak.array_equal(a, returned)
-
-
-@settings(max_examples=500)
-@given(a=st_ak.constructors.arrays())
-def test_roundtrip_no_equality_check(a: ak.Array) -> None:
-    """`to_buffers` followed by `from_buffers` does not error.
-
-    This test includes all array types (including union arrays)
-    but does not assert equality because `ak.array_equal`
-    returns incorrect results for some union arrays.
-    See `test_roundtrip` for equality assertions on the supported subset.
-    """
-    sent = ak.to_buffers(a)
-    ak.from_buffers(*sent)

